tests/unit/forklift/test_legacy.py

```diff
@@ -429,6 +453,41 @@ def test_bails_with_invalid_zipfile(self, tmpdir, filename, filetype):
 
         assert not legacy._is_valid_dist_file(f, filetype)
 
+    @pytest.mark.parametrize(
+        "filename", ["test.tar", "test.tar.gz", "test.tgz", "test.tar.bz2", "test.tbz2"]
+    )
+    def test_bails_with_invalid_tarfile(self, tmpdir, filename):
+        fake_tar = str(tmpdir.join(filename))
+
+        with open(fake_tar, "wb") as fp:
+            fp.write(b"Definitely not a valid tar file.")
+
+        assert not legacy._is_valid_dist_file(fake_tar, "sdist")
+
+    @pytest.mark.parametrize("compression", ("", "gz", "bz2"))
+    def test_tarfile_validation(self, tmpdir, compression):
+        file_extension = f".{compression}" if compression else ""
+        tar_fn = str(tmpdir.join(f"test.tar{file_extension}"))
+        data_file = str(tmpdir.join("dummy_data"))
+
+        with open(data_file, "wb") as fp:
+            fp.write(b"Dummy data file.")
+
+        with tarfile.open(tar_fn, "w:" + compression) as tar:
+            tar.add(data_file, arcname="package/module.py")
+
+        assert not legacy._is_valid_dist_file(
+            tar_fn, "sdist"
+        ), "no PKG-INFO; should fail"
+
+        os.unlink(tar_fn)
```

Comment:
Instead of unlinking `tar_fn` here, wouldn't it be better to make this a separate test?

Reply:
good idea. done.